### DISCUSSION TITLE
Docker - How is ".serverless" dir path selected (and why is it '/.serverless')

### DISCUSSION THREAD

**Discussion by Fran-Rg** [serverless/serverless#10809] [Q&A] [answered]
## Issue

I'm trying to understand why serverless is trying to create a folder a the root of my environment and not in the current folder
```
sls package --verbose
Running "serverless" from node_modules
Environment: linux, node 17.6.0, framework 3.7.1 (local) 3.7.1v (global), plugin 6.1.5, SDK 4.3.2
Docs:        docs.serverless.com
Support:     forum.serverless.com
Bugs:        github.com/serverless/serverless/issues

Error:
Error: EACCES: permission denied, mkdir '/.serverless'

sh-4.2$ sls package --verbose --package ${PWD}/.here
Running "serverless" from node_modules
Environment: linux, node 17.6.0, framework 3.7.1 (local) 3.7.1v (global), plugin 6.1.5, SDK 4.3.2
Docs:        docs.serverless.com
Support:     forum.serverless.com
Bugs:        github.com/serverless/serverless/issues

Error:
Error: EACCES: permission denied, mkdir '/.serverless'
```

## Context

I'm running in a docker image with a non sudo user. It has serverless installed globally. When running sls package the problem happens. 
I've tried to use the --package option but it's not working as I would expect: https://www.serverless.com/framework/docs/providers/aws/guide/packaging

The user is not admin for safety:
```
whoami
whoami: cannot find name for user ID 1000
```
But the user is allowed to create the directory in the current path:
```
pwd
/Users/XYZ/work/lambda-layers-shared/layers/requirements-shared
mkdir .serverless
```
This works

The serverless.yaml file:
```
service: requirements-shared

custom:
  pythonRequirements:
    pythonBin: python3
    layer:
      name: ${self:service}
      description: Python requirements lambda layer
      compatibleRuntimes:
        - python3.8

frameworkVersion: ">=2.0.0"
provider:
  name: aws
  region: eu-west-1

plugins:
  - serverless-python-requirements
```

## Bonus points
* Why does the --verbose option doesn't provide more information?
* Why is there no stacktrace?

**Comment by pgrzesik** [comment]
Hello @Fran-Rg - the `.serverless` directory is used for a few things:
- it stores/caches artifacts used for custom resources
- it stores/caches compiled schema validator for ajv
- it stores data for autocompletion functionality
- it stores telemetry data
- it can also store binary installation of the Framework

So as you can see, it actually serves quite a few purposes and it will always be created when it's needed for the first time.

**Reply by Fran-Rg** [reply]
Hi, thanks the answer but I think you misunderstood a little bit my question. You answered "what is the .serverless directory for", my question was "how does serverless decide where to create it? and how to change that"

Inside a docker running as non sudo it's trying to create the ".serverless" dir at the root (e.g. `/.serverless`). Why did it take this decision when my serverless.yaml is at "`/home/project/XYZ/serverless.yam`". Since the user isn't root, `mkdir /.serverless` fails

**Reply by pgrzesik** [reply]
Ah, sorry @Fran-Rg - it's creating this folder in the home directory - so it seems like in your case the home directory is `/`. It does not create it in the context of your project directory because the contents on it are supposed to be reused across different projects. Actually, other `.serverless` directory is also created in your service directory, but it stores the files needed for deployment of the specific project.

**Reply by Fran-Rg** [reply]
And is there any option to control this path to not use the home directory? Or shall I `export HOME=${PWD}`?

**Reply by pgrzesik** [reply]
There is no option to control this path at the moment - changing `HOME` to your service directory will probably break - I would really not recommend it - as I mentioned, in your service directory, there will also be a `.serverless` dir but that one serves a totally different purpose.

**Reply by Fran-Rg** [reply]
oh, very interesting and confusing, 2 `.serverless` directories with different purpose. I'll get a better place to set this home then without using the service directory
After a quick test that got me sorted

**Reply by pgrzesik** [reply]
Setting HOME to a different directory than service directory should do the trick in your case 👍

**Comment by Fran-Rg** [comment, marked as answer]
Just to tldr above
2 .serverless directories get created one in the service and one in the home directory. Having HOME set to '/' was causing the issue.

Since it's docker, Setting HOME to a different directory than service directory did the trick in this case
